doc/design/block.md

```diff
@@ -0,0 +1,167 @@
+
+RNNOp looks like the loop structure in programming languages.
+And similarly, WhileOp and IfElseOp are like loop and conditions respectively.
+So we want to verify if we should have a class Block in PaddlePaddle that works like a pair of curly braces in the loop and condition structures of programming languages.
+
+Blocks do not only group source code, but also narrow the lexical scope of variables so that they do not conflict with variables having the same name used elsewhere in a program.
+
+In Paddle, we need a similar concept called Block to support following scenes:
+
+- define a PaddlePaddle program by writing blocks of codes, which includes the definitions of variables and operators.
+  - `RNNOp`, `SwitchOp`, `WhileOp` and `IfElseOp` needs Block to help to define sub-block. 
+- help to execute multiple operators, blocks should group operators and runs like a single operator.
+
+## How to use Block
+In `RNNOp`, `SwitchOp`, `WhileOp` and `IfElseOp`, a sub-block should be used to help to define a sub-block.
+
+Let's start from how a RNNOp is described using Block:
+
+```python
+v = some_op()
+m_boot = some_op()
+
+W = pd.Variable(shape=[20, 20])
+U = pd.Varable(shape=[20, 20])
+
+rnn = create_RNNOp()
+with rnn.stepnet() as net:
+  # declare the input variables that need to be segmented into steps
+  x = net.add_input(v)
+  # declare rnn's memory (state)
+  h = net.add_memory(init=m_boot)
+  
+  fc_out = pd.matmul(W, x)
+  hidden_out = pd.matmul(U, h)
+  sum = pd.add_two(fc_out, hidden_out)
+  act = pd.sigmoid(sum)
+  # declare outputs that needs to be merged across all the steps
+  net.add_output(act, hidden_out)
+
+acts, hs = rnn()
+```
+
+Above Python program builds a protobuf message which describe the model, a C++ Block class will create the corresponding Variables and Operators, and execute all the operators.
+
+## Block Implementation
+
+During the Protobuf message generation, the Block should store VarDesc (the Protobuf message which describes Variable) and OpDesc (the Protobuf message which describes Operator).
+
+VarDesc in a block should have its name scope to avoid local variables affect father block's name scope. 
+Child block's name scopes should inherit the father's so that OpDesc in child block can reference a VarDesc that stored in father block. For example
+
+```python
+a = pd.Varaible(shape=[20, 20])
+b = pd.fc(a, params=["fc.w", "fc.b"])
+
+rnn = pd.create_RNNOp()
+with rnn.stepnet() as net:
+    x = net.add_input(a)
+    # reuse fc's parameter
+    fc_without_b = pd.get_variable("fc.w")
+    net.add_output(fc_without_b)
+
+out = rnn()
+```
+the method `pd.get_variable` can help retrieve a Variable by a name, a Variable may store in a father block, but might be retrieved in a child block, so block should have a variable scope that supports inheritance.
+
+We can implement this idea in the following approach
+
+```c++
+class Block {
```

Comment:
If I was not misunderstanding, `Block` is used to construct Protobuf not for running?

Reply:
The current design of `Block` mix the compile time `Block` (which is used for constructing the computation graph of Block) and runtime `Block` (which is used for executing the computation graph) in the same class `Block`.

I think it is cleaner to separate them into two classes, `BlockCompiler` or `BlockOp`.

```cpp
class BlockCompiler {
 public:
  OpDesc* AppendOp();
  // Compile and generate new runtime block op
  BlockOp* Compile() const;
 private:
  BlockCompiler* parent;
  map<...> var_desc_;
  map<...> op_desc_;
};

class BlockOp : public OperatorBase {
 public:
  void Run(...) const;
  void InferShape(...) const;
 private:
  ...
};
```

Reply:
good idea

Reply:
Or maybe `BlockCompiler` is a bad name, we could consider a better name. Maybe it could be [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree) or some name related to compiler.